### DISCUSSION TITLE
Worker-side hooking into log generation

### DISCUSSION THREAD

**Discussion by vpax** [zeek/zeek#1972] [Ideas]
Currently the user can hook into log generation via the `Log::log_stream_policy` hook, but this happens on the Logger node.  It could be very handy to provide a means to hook instead on the worker generating a given log entry, as a way of spreading out load.  I think we could do this by adding a hook just before:
https://github.com/zeek/zeek/blob/d77fd8305d68a00330e8c0740390f22d82b2e2ac/scripts/base/frameworks/logging/main.zeek#L798
Seems simple enough, and in the common case of no hook being provided, pretty cheap.

Thoughts?

**Comment by awelzel** [comment]
> Thoughts?
> Currently the user can hook into log generation via the Log::log_stream_policy hook, but this happens on the Logger node.

Hmm, log_stream_policy is handled on the worker node, just like `$ev` with the original "Info" record. Dispatching to the writers which than hands it off to the logger nodes happens later. So it may already do what you need?

https://github.com/zeek/zeek/blob/master/src/logging/Manager.cc#L726

One nice feature of the  `Log::log_stream_policy` hook is that it allows to tap into the logging stream and bypass the existing log writer functionality to instead implement it via scripting.

**Reply by vpax** [reply]
Oh!  Well then that's easy :-).  I had noticed that the hook isn't called by scripting but instead inside the event engine from `logging/Manager.cc`. Given that unusual design I assumed it was the logger doing so rather than the worker, good to hear that it isn't.  (It does make me wonder about why it's done that way - my guess is because the use of the call-out predates the introduction of hooks.)

**Comment by rsmmr** [comment]
No, it's because all the script-level log processing happens on the workers (more precisely: wherever the log entry is being created). The content of a log line is fully finalized at the originating node, and then sent over to the logger as a binary serialization that's directly written out over there once received. That's for performance: this way, the log entries don't need to go through scriptland at the logger node, which would create a bottleneck there and limit scalability. (So, it sounds like, for exactly the reason you're asking this question in the first place) :-) )

**Comment by vpax** [comment]
Cool, that means we can close this out.
